Ical.Net/CollectionExtensions.cs

```diff
@@ -0,0 +1,58 @@
+﻿//
+// Copyright ical.net project maintainers and contributors.
+// Licensed under the MIT license.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Ical.Net.CalendarComponents;
+using Ical.Net.DataTypes;
+using Ical.Net.Evaluation;
+
+namespace Ical.Net;
+
+public static class CollectionExtensions
+{
+    /// <summary>
+    /// Returns the elements of the sequence of occurrences to only include those that start before the specified period end.
+    /// Important: The input sequence <b>must be ordered</b> according to the type's default equality comparer.
+    /// <para/>
+    /// A perfect fit is the sequence returned by <see cref="RecurringComponent.GetOccurrences(CalDateTime, EvaluationOptions)"/>,
+    /// like the frequently used <see cref="CalendarEvent.GetOccurrences(CalDateTime, EvaluationOptions)"/>
+    /// or <see cref="Calendar.GetOccurrences"/>.
+    /// <para/>
+    /// This is a convenience method for <see cref="Enumerable.TakeWhile{TSource}(IEnumerable{TSource}, Func{TSource, bool})"/>.
+    /// But even with this method, the input sequence must be ordered according to the type's default equality comparer.
+    /// </summary>
+    /// <param name="sequence">The ordered sequence to be filtered.</param>
+    /// <param name="periodEnd">The exclusive end of the period to be used as a filter.</param>
+    /// <returns>
+    /// The elements of the sequence of occurrences to only include those that start before the specified period end.
+    /// </returns>
+    public static IEnumerable<Occurrence> TakeBefore(this IEnumerable<Occurrence> sequence, CalDateTime? periodEnd)
+        => (periodEnd == null) ? sequence : sequence.TakeWhile(p => p.Period.StartTime < periodEnd);
```

Comment:
I think `periodEnd` shouldn't be nullable. In the test space I made it nullable to act as a c&p replacement for the previous periodEnd param and its used just internally (actually this was part of the reason not making it public). But in this case I feel its not quite natural allowing to pass a null value, making it a noop. I therefore suggest making it non-null. In the test projects we could keep an internal extension method that still allows passing null for simplicity.

Reply:
Actually this was my first thought as well.: `null` would be a noop. Just like "abc".ToString()` returning `this`.
However, users could then use `toDate` e.g. in a loop, without caring for nullability. See our own tests below.

If `toDate` was **not** nullable, we could (would) write the unit test like
```cs
var occurrences = toDate != null
    ? evt.GetOccurrences(fromDate).TakeBefore(toDate).ToList()
    : evt.GetOccurrences(fromDate).ToList();
```
without the need for an extension method just for this test.
What's better?

Reply:
I'd prefer the not nullable, just like .TakeWhile or .Skip don't accept null. For our own use we can add the nullable version, like everyone can easily do. Outside of test cases I wouldn't expect it to be a common case that invocations need the nullable version. Either they want to limit or they don't I'd expect.

Reply:
Then so be it.